iwyu_ast_util.cc

```diff
@@ -1266,20 +1266,33 @@ const Type* TypeOfParentIfMethod(const CallExpr* expr) {
 }
 
 const Expr* GetFirstClassArgument(CallExpr* expr) {
-  for (CallExpr::arg_iterator it = expr->arg_begin();
-       it != expr->arg_end(); ++it) {
-    const Type* argtype = GetTypeOf(*it);
-    // Make sure we do the right thing given a function like
-    //    template <typename T> void operator>>(const T& x, ostream& os);
-    // In this case ('myclass >> os'), we want to be returning the
-    // type of os, not of myclass, and we do, because myclass will be
-    // a SubstTemplateTypeParmType, not a RecordType.
-    if (isa<SubstTemplateTypeParmType>(argtype))
-      continue;
-    argtype = argtype->getUnqualifiedDesugaredType();  // see through typedefs
-    if (isa<RecordType>(argtype) ||
-        isa<TemplateSpecializationType>(argtype)) {
-      return *it;
+  if (const FunctionDecl* callee_decl = expr->getDirectCallee()) {
+    if (isa<CXXMethodDecl>(callee_decl)) {
+      // If a method is called, return 'this'.
+      return expr->getArg(0);
+    }
+    // Handle free functions.
+    int params_count = callee_decl->getNumParams();
```

Comment:
... and actually, is that a strong invariant? I would expect things like:

    void f(int a, int b=0);

    f(10);

to break that.

Can't think of a case with operators and default arguments, so not sure if it's relevant.